Lombiq.Hosting.MediaTheme.Deployer/Program.cs

```diff
@@ -49,102 +82,158 @@ private static void HandleParseError(IEnumerable<Error> errors)
         }
     }
 
-    private static void RunOptions(CommandLineOptions values)
+    private static async Task RunOptionsAsync(CommandLineOptions options)
+    {
+        try
+        {
+            await RunOptionsInnerAsync(options);
+        }
+        catch (Exception ex)
+        {
+            WriteLine("Deployment failed with the following exception: {0}", ex.ToString());
+            Environment.ExitCode = 1;
+        }
+    }
+
+    [SuppressMessage(
+        "Major Code Smell",
+        "S4457:Parameter validation in \"sync\"/\"await\" methods should be wrapped",
+        Justification = "RunOptionsAsync() needs to use await as well to be able to set the exit code on exception.")]
+    private static async Task RunOptionsInnerAsync(CommandLineOptions options)
     {
         // Creating directory for the deployment.
-        var newDirectoryPath = CreateNewDirectoryPath(values);
+        var newDirectoryPath = CreateNewDirectoryPath(options);
 
         try
         {
-            // Determine whether the directory exists.
-            if (Directory.Exists(newDirectoryPath))
-            {
-                WriteLine("That directory already exists.");
-                return;
-            }
+            if (Directory.Exists(newDirectoryPath)) Directory.Delete(newDirectoryPath, recursive: true);
 
-            // Try to create the directory.
             Directory.CreateDirectory(newDirectoryPath);
 
-            WriteLine("The directory was created successfully. {0}", newDirectoryPath);
+            WriteLine("The \"{0}\" directory was created successfully.", newDirectoryPath);
         }
-        catch (Exception exception)
+        catch (Exception)
         {
-            WriteLine("The directory creation failed: {0}", exception.ToString());
-            return;
+            WriteLine("Creating the directory {0} failed.", newDirectoryPath);
+            throw;
+        }
+
+        var themePath = options.ThemePath;
+
+        if (string.IsNullOrEmpty(themePath))
+        {
+            throw new ArgumentException("The theme's path must be provided.");
         }
 
-        var pathToTheme = values.PathOfTheTheme;
+        var recipeSteps = new JArray();
+
+        // Creating Feature step to enable the Media Theme theme and Bridge module.
+        dynamic featureStep = new JObject();
+        featureStep.name = "Feature";
+        featureStep.enable = new JArray("Lombiq.Hosting.MediaTheme.Bridge", "Lombiq.Hosting.MediaTheme");
+        recipeSteps.Add(featureStep);
```

Comment:
dynamic 🤮🤮🤮

Could you replace this and the others below with stuff like this?
```csharp
        var featureStep = JObject.FromObject(new
        {
            name = "Feature",
            enable = new[] { "Lombiq.Hosting.MediaTheme.Bridge", "Lombiq.Hosting.MediaTheme" },
        });
```
Or better yet, the `recipeSteps` could be a `List<object>` and then it'll get converted when it's passed to its holder object in `CreateRecipeAndWriteIt()`. Personally I'd replace the whole thing with POCOs, but at least we should get rid of that cursed keyword.

Reply:
Be not afraid!

I'm not sure if it was worth it, but I got rid of `dynamic`. I don't see why we'd use `List<object>` as the parameter in `CreateRecipeAndWriteIt()`, since we're building a piece of JSON, so using `JArray` and friends is adequate.

Reply:
> I'm not sure if it was worth it

It's always worth it, `dynamic` is from the devil.

> I don't see why we'd use List<object> as the parameter

All you do with `recipeSteps` is pass it into `CreateRecipeAndWriteIt()` where it becomes a property of an object that gets converted via its `JObject.FromObject()` anyway. So it's not actually necessary to convert the individual items one by one.

Reply:
I'd like to believe `JObject.FromObject()` doesn't need to work on converting a `JArray` to a `JArray` :). `List<object>` is really a catch-all, as mentioned I see the semantics of `JArray` more appropriate, and in practice there's no difference.

Reply:
My point is about code simplicity, not performance. For example I figure it's better to do
`JObject.From( { array = new List<object> { item1, item2, item3 } } )` than
`JObject.From( { array = new JArray() { JObject.From(item1), JObject.From(item2), JObject.From(item3) } } )`
I don't think littering the code with JSON-specific abstraction is helpful when you are just building an object.